src/dc_tools.rs

```diff
@@ -1476,6 +1476,111 @@ pub unsafe fn dc_make_rel_and_copy(context: &Context, path: *mut *mut libc::c_ch
     success
 }
 
+#[derive(Debug, PartialEq)]
+pub enum CStringError {
+    InteriorNullByte,
+    NotUnicode,
+}
+
+impl std::fmt::Display for CStringError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            CStringError::InteriorNullByte => write!(f, "String contains an interior null byte"),
+            CStringError::NotUnicode => write!(f, "String is not valid unicode"),
+        }
+    }
+}
```

Comment:
we should have the `failure` crate in our toolbox, making this error generation a bi tnicer

```rust
#[derive(Debug, Fail)]
enum CStringError {
    #[fail(display = "String contains an interior null bytes")
    InvalidToolchainName
}
```

Reply:
also you should implement the `Error` trait if you write your own error types, to be a nice rust citizien

Reply:
Agreed on failure, just wanted to start without adding this thing.

Didn't realise I needed to do anything for the `Error` trait, thought everything was implemented.  Anyway, I guess switching to failure::Fail makes this go away as it does that automatically?

Reply:
> makes this go away as it does that automatically?
yes it does

Reply:
Now using failure.